canary/producer-cpp/src/Canary.cpp

```diff
@@ -0,0 +1,865 @@
+                Aws::CloudWatch::Model::PutMetricDataRequest cwRequest;
+                cwRequest.SetNamespace("KinesisVideoSDKCanary");
+
+                auto currentTimestamp = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
+                auto persistedAckLatency = (currentTimestamp - timeOfFragmentEndSent); // [milliseconds]
+                pushMetric("PersistedAckLatency", persistedAckLatency, Aws::CloudWatch::Model::StandardUnit::Milliseconds, persistedAckLatencyDatum, data->pDimensionPerStream, cwRequest);
+                LOG_DEBUG("Persisted Ack Latency: " << persistedAckLatency);
+                if (data->pCanaryConfig->useAggMetrics)
+                {
+                    pushMetric("PersistedAckLatency", persistedAckLatency, Aws::CloudWatch::Model::StandardUnit::Milliseconds, persistedAckLatencyDatum, data->pAggregatedDimension, cwRequest);
+
+                }
+                data->pCWclient->PutMetricDataAsync(cwRequest, onPutMetricDataResponseReceivedHandler);
+                break;
+            }
+            case FRAGMENT_ACK_TYPE_RECEIVED:
+            {
+                Aws::CloudWatch::Model::MetricDatum receivedAckLatencyDatum;
+                Aws::CloudWatch::Model::PutMetricDataRequest cwRequest;
+                cwRequest.SetNamespace("KinesisVideoSDKCanary");
+
+                auto currentTimestamp = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
+                auto receivedAckLatency = (currentTimestamp - timeOfFragmentEndSent); // [milliseconds]
+                pushMetric("ReceivedAckLatency", receivedAckLatency, Aws::CloudWatch::Model::StandardUnit::Milliseconds, receivedAckLatencyDatum, data->pDimensionPerStream, cwRequest);
+                LOG_DEBUG("Received Ack Latency: " << receivedAckLatency);
+                if (data->pCanaryConfig->useAggMetrics)
+                {
+                    pushMetric("ReceivedAckLatency", receivedAckLatency, Aws::CloudWatch::Model::StandardUnit::Milliseconds, receivedAckLatencyDatum, data->pAggregatedDimension, cwRequest);
+                }
+                data->pCWclient->PutMetricDataAsync(cwRequest, onPutMetricDataResponseReceivedHandler);
+                break;
+            }
+            case FRAGMENT_ACK_TYPE_BUFFERING:
+            {
+                cout << "FRAGMENT_ACK_TYPE_BUFFERING callback invoked" << endl;
+                break;
+            }
+            case FRAGMENT_ACK_TYPE_ERROR:
+            {
+                cout << "FRAGMENT_ACK_TYPE_ERROR callback invoked" << endl;
+                break;
+            }
+        }
+    }
+}
+
+}  // namespace video
+}  // namespace kinesis
+}  // namespace amazonaws
+}  // namespace com;
+
+// add frame pts, frame index, original frame size, CRC to beginning of buffer
+VOID addCanaryMetadataToFrameData(PFrame pFrame)
+{
+    PBYTE pBuf = new BYTE[CANARY_METADATA_SIZE + pFrame->size];
```

Comment:
You can avoid this by just memcpy'ing in `create_kinesis_video_frame`. frameData is being reinterpreted there. All you need to do is `MEMCPY(frame->frameData + METADATA_SIZE, frame->frameData, len) and set `frame->size` to include `METADATA_SIZE` and it will be handled. Then invoking this function is as it is done here. Your implementation would lead to memory leak because you do not clear the memory.